### DISCUSSION TITLE
check the service status (stopped or running)

### DISCUSSION THREAD

**Discussion by roxxcorp** [autokey/autokey#750] [Q&A] [answered]
hello,

 I need sometimes to pause the service but Is there a way to retrieve the service status.

I would be able to have a system tray «on/off light» to remind me instead of having to try to copy paste in an editor to see if my shortcuts are french azerty or french bepo.

Thanks.

Antoine.

**Comment by josephj11** [comment]
TL;DR: AFAIK, there's no easy way to do this. However, if you're really motivated, you can do it yourself.

There are two questions here: Disabling AutoKey and getting its status.

I don't know of an elegant way to enable and disable monitoring, but if you want this bad enough, I see two ways to approach it.

The simplest approach is to carefully think through exactly how you would like this feature to work and file an enhancement issue requesting it.

The problem with this approach is that we have extremely limited developer resources at the moment, so it might never get implemented.

The more difficult approach would be to do it yourself in an AutoKey script. Without the relevant API calls, you would have to manipulate the GUI itself which is a bit of work.

Here's roughly how I would do it:

All of this relies on using [visgrep() ](http://autokey.github.io/autokey/api/highlevel.html) which makes use of the Xautomation project.

You manually create small region screenshots of the features on the screen you want to locate and tell visgrep() to find them for you. It returns the screen coordinates where it found a match or throws an exception for not found.

The first thing you search for would be the AutoKey icon in your system tray. Once found, use its coordinates to right click the mouse on it. This will display the AutoKey icon context menu.

The next thing you would search for is (an image of) the checkbox to the left of Enable monitoring in that menu.
If the image is of a filled in checkbox, then monitoring is enabled. If the checkbox is empty, then monitoring is disabled.
You can then use the mouse to click on the checkbox to toggle its status as desired.

Your script could also create/delete a semaphore file that only exists when monitoring is enabled.

You could then write a (daemon) script (in bash or in the language of your choice) that would periodically look for the semaphore file and do something based on its existence. This gets a bit interesting. I currently have a bash script that adds/deletes an icon on my desktop based on testing a particular condition (not AutoKey related). That's easy enough, but I'm not sure how I would do something like that in my system tray. I do know it's possible because applications such as AutoKey and Dropbox have icons in the panel which change to reflect certain conditions.

If you decide to pursue this, come over to [Gitter](https://gitter.im/autokey/autokey) and we can help you work through it.

**Comment by josephj11** [comment, marked as answer]
There may be a much simpler way to approach this.

Instead of enabling/disabling AutoKey, you could make a set of your scripts run conditionally. This will only work for scripts, but writing scripts to emulate phrases is trivial.

Each script in this set would start by fetching an AutoKey stored global variable. If the variable is set to some value signifying "off", the script would just emit the hotkey that triggered it, making it effectively transparent. If the value signifies "on", then the script would proceed to do its work. 

You would then have another script(s) which set (toggle) the AutoKey stored global variable to the appropriate value signifying "on" or "off".

Running this script via a hotkey would modify the behavior of all the other scripts in this set.

This approach has the additional advantage of leaving AutoKey active so that other unrelated actions are still available to be used.

This still doesn't put an indicator in your panel.

A workaround for that might be an AutoKey script that displays an informational pop-up displaying the status of the global stored variable.

A refinement of this would be explicitly calling out to the dialog manager directly. This would give you (among other things) the option of specifying a timeout for the dialog so that you don't have to manually dismiss it.

**Reply by roxxcorp** [reply]
Hi and thank you for the quick answer.

I decided to go for your second approach. I don't know if I did it properly but this how I managed to make it work :+1:

### 1 : I added a new global variable in .config/autokey/autokey.json
```
"scriptGlobals": {
    "state": 1
},
```

### 2 : As you said, create a script in autokey to retrieve and toggle this variable using a hotkey :
```
# Enter script code
print("0-state:", store.get_global_value("state"))

output = ""
if store.get_global_value("state") == 1:
    store.set_global_value("state", 0)
    output = system.exec_command("polybar-msg action autokey hook 1", getOutput=True)
elif store.get_global_value("state") == 0:
    store.set_global_value("state", 1)
    output = system.exec_command("polybar-msg action autokey hook 0", getOutput=True)

print(output)
print("1-state:", store.get_global_value("state"))

```
### 3 : Before each of my shortcuts, I added a check for the global state value like this :
```
if store.get_global_value("state") == 1:
    keyboard.send_keys("<ctrl>+a")

```

**(AND FOR THOSE WHO ARE INTERESTED…)**
### 4 : As you see, I use polybar and the «ipc» messages to tell one of the polybar module to change an icon in my menubar :
```
output = system.exec_command("polybar-msg action autokey hook 1", getOutput=True)
output = system.exec_command("polybar-msg action autokey hook 0", getOutput=True)
```

### 5: In my polybar config file, I put this module:
```
[module/autokey]
type = custom/ipc
hook-0= echo -ne "\\uf058"
hook-1= echo -ne "\\uf111"
initial = 1
```
When autokey send the ipc message to polybar, it changes a little icon in my bar :
french bepo shortcuts on :
![yes](https://user-images.githubusercontent.com/47446975/205446076-4784f8ec-1d46-4c55-896c-9350f29bdde2.jpg)

french bepo shortcuts off:
![no](https://user-images.githubusercontent.com/47446975/205446116-06af4a2e-4014-4f6a-b091-25597387f535.jpg)


**And really thank you for the help !!!**

Antoine.

**Reply by josephj11** [reply]
Have to look at polybar. Sounds useful. Glad it works for you.

It sounds like you edited autokey.json manually. If it works, fine, but our [API](http://autokey.github.io/autokey/api/store.html#autokey.scripting.Store.set_global_value) lets you create global variables and that should survive any changes we might make in the future.

If a stored value is retrieved before it is initially defined, I think it throws an exception, but you can handle that with a try: block.

After you've used your new code for a while, you might want to consider adding it as an example in our wiki.

As you probably already know, print() (for debugging, etc.) only works if the active window is a terminal or something else  that uses stdout/stdin. You can always use our dialog API calls to display things in the GUI or run an external dialog manager from an AutoKey script.

**Reply by roxxcorp** [reply]
> Have to look at polybar. Sounds useful. Glad it works for you.


Yes. Simple to configure and efficient.

> It sounds like you edited autokey.json manually. If it works, fine, but our [API](http://autokey.github.io/autokey/api/store.html#autokey.scripting.Store.set_global_value) lets you create global variables and that should survive any changes we might make in the future.

It works when I use autokey.json. I wanted to use the scipting.store way of doing it but I don't know how to start a script at the the startup of the application. I only know to launch a script by using a shortcuts or a hotkey. If you have the answer, I will give it a try.

> After you've used your new code for a while, you might want to consider adding it as an example in our wiki.

Ok, if everything works, I will put it in the wiki. Maybe I will try to put it in the autokey app but I'm not a python expert but it could be a good training. :-)

> As you probably already know, print() (for debugging, etc.) only works if the active window is a terminal or something else that uses stdout/stdin. You can always use our dialog API calls to display things in the GUI or run an external dialog manager from an AutoKey script.

Yes, print() was just used for debugging purpose.

**Reply by josephj11** [reply]
I have never used it, but that's what autokey-run is for. As long as AutoKey is already running, you can wrap your application startup command in a (bash...) script that runs autokey-run to get AutoKey ready for it.

I encourage you to join us over on [Gitter](https://gitter.im/autokey/autokey) where the devs and advanced users hang out. That's where you can get guidance on autokey-run and any questions you might have about creating a PR to add to AutoKey. I don't think there is any documentation, etc. for autokey-run other than its very brief man page.
